### DISCUSSION TITLE
Error connecting MQTT: module 'paho.mqtt.client' has no attribute 'CallbackAPIVersion'

### DISCUSSION THREAD

**Discussion by joevoo82** [philippoo66/optolink-splitter#60] [Q&A] [answered]
Ich hatte bis gestern probeweise alles am Laufen, allerdings ohne VitoConnect.
Nachdem heute der CP2102 eingebaut wurde, trete ich mich von Mine zu Mine und komm nun nicht mehr weiter.

Beim Start in der venv bekomme ich folgende Fehlermeldung: `Error connecting MQTT: module 'paho.mqtt.client' has no attribute 'CallbackAPIVersion'`

Paho ist installiert: `python3-paho-mqtt/stable,stable,now 1.6.1-1 all [installed]`

Brauche ich hier schon die 2.0.0 oder höher? Ich bin auf nem RPi 4 mit Bookworm unterwegs.

```
python-apt-common/stable,stable,now 2.6.0 all [installed,automatic]
python-is-python3/stable,stable,now 3.11.2-1+deb12u1 all [installed]
python3-apt/stable,now 2.6.0 arm64 [installed,automatic]
python3-certifi/stable,stable,now 2022.9.24-1 all [installed,automatic]
python3-chardet/stable,stable,now 5.1.0+dfsg-2 all [installed,automatic]
python3-charset-normalizer/stable,stable,now 3.0.1-2 all [installed,automatic]
python3-colorzero/stable,stable,now 2.0-2 all [installed,automatic]
python3-debconf/stable,stable,now 1.5.82 all [installed,automatic]
python3-distro/stable,stable,now 1.8.0-1 all [installed,automatic]
python3-distutils/stable,stable,now 3.11.2-3 all [installed,automatic]
python3-gpiozero/stable,stable,now 2.0.1-0+rpt1 all [installed]
python3-idna/stable,stable,now 3.3-1+deb12u1 all [installed,automatic]
python3-lgpio/stable,now 0.2.2-1~rpt1 arm64 [installed,automatic]
python3-lib2to3/stable,stable,now 3.11.2-3 all [installed,automatic]
python3-libgpiod/stable,now 1.6.3-1+b3 arm64 [installed]
python3-minimal/stable,now 3.11.2-1+b1 arm64 [installed,automatic]
python3-paho-mqtt/stable,stable,now 1.6.1-1 all [installed]
python3-pigpio/stable,stable,now 1.79-1+rpt1 all [installed]
python3-pip-whl/stable,stable,now 23.0.1+dfsg-1+rpt1 all [installed,automatic]
python3-pkg-resources/stable,stable,now 66.1.1-1 all [installed,automatic]
python3-pycryptodome/stable,now 3.11.0+dfsg1-4 arm64 [installed,automatic]
python3-requests/stable,stable,now 2.28.1+dfsg-1 all [installed,automatic]
python3-rpi-lgpio/stable,stable,now 0.6-0~rpt1 all [installed]
python3-serial/stable,stable,now 3.5-1.1 all [installed]
python3-setuptools-whl/stable,stable,now 66.1.1-1 all [installed,automatic]
python3-six/stable,stable,now 1.16.0-4 all [installed,automatic]
python3-smbus2/stable,now 0.4.2-1 arm64 [installed]
python3-spidev/stable,now 20200602~200721-1+bookworm arm64 [installed]
python3-toml/stable,stable,now 0.10.2-1 all [installed,automatic]
python3-urllib3/stable,stable,now 1.26.12-1 all [installed,automatic]
python3-venv/stable,now 3.11.2-1+b1 arm64 [installed]
python3.11-minimal/stable,now 3.11.2-6+deb12u4 arm64 [installed,automatic]
python3.11-venv/stable,now 3.11.2-6+deb12u4 arm64 [installed,automatic]
python3.11/stable,now 3.11.2-6+deb12u4 arm64 [installed,automatic]
python3/stable,now 3.11.2-1+b1 arm64 [installed]
```
Danke für die Hilfe!

**Comment by philippoo66** [comment]
moin @joevoo82 !

das ist erstmal verwunderlich. das CP2102 geht ja auf eine UART des Raspi und hat rein garnichts mit dem MQTT client zu tun. Wenn MQTT vorher funktioniert hat, sollte es nachher genauso funktionieren.

ich hab grad auf meinem 'Produktiv-System' geschaut, da ist allerdings auch Ver. 2 installiert:
```
pi@raspberrypi2:~ $ pip show paho-mqtt
Name: paho-mqtt
Version: 2.0.0
Summary: MQTT version 5.0/3.1.1 client class
Home-page: None
Author: None
Author-email: Roger Light <roger@atchoo.org>
License: EPL-2.0 OR BSD-3-Clause
Location: /usr/local/lib/python3.9/dist-packages
Requires:
Required-by:
```

Aber wenn du das Ganze eh in einer venv am laufen hast - was hindert dich da dran _in der venv_(!) Vers. 2 zu installieren? venv aktivieren und dann pip install phao-mqtt. Die ist dann nur in deiner venv und das hat dann keinen Einfluss auf irgendwas andres auf deinem Raspi.

Wie kommt die alte Version da überhaupt rein? die ist 3 Jahre alt, so lange gibt es den Splitter glaubich noch garnicht. Selbst Bookworm kam erst Mitte 2023...

**Reply by philippoo66** [reply]
ps.
"Paho-MQTT version 2.0.0 for Python was officially released on October 26, 2021. This version introduced breaking changes, such as **modifications to the callback API versioning**, and improvements in documentation. For migration and detailed updates, the release notes and changelog are available on GitHub and the Eclipse Paho project page"

**Comment by joevoo82** [comment, marked as answer]
Moin @philippoo66 ! Vielen Dank für die Hilfe, das wars! Warum im root paho in der 1.6. installiert wird, mag verstehen der will. Aber mit deinem Hinweis auf die venv, wird dort nach pip install tatsächlich die 2.x installiert. Es läuft, ich bin begeistert! Vielen Dank für dieses tolle Projekt!

**Reply by philippoo66** [reply]
gern, freut mich dass es dir gefällt! danke für die Rückmeldung.

Grüsse!
Phil
